### PR TITLE
fix for New spatial mapping drops frames every update period 

### LINKED ISSUE #207: SpatialUnderstanding.cs throws warning about unreachable code
`Assets\HoloToolkit\SpatialUnderstanding\Scripts\SpatialUnderstanding.cs(53,17): warning CS0162: Unreachable code detected`

### LINKED ISSUE #264: New spatial mapping drops frames every update period
The new spatial mapping mesh code is dropping several frames every update period.  This makes the mesh blink on and off in an unsettling manner.   

I have tracked it down to the fact that 'yields' are being called between the mesh 'Cleanup()' and the completion of the 'AddSurfaceObject' calls.   During this time, the meshes are in an inconsistent state.

From line 252 of SpatialUnderstandingCustomMesh.cs: I tried changing it from:

```
252             yield return null;

                // Clear our old surface objects.
                Cleanup();

                yield return null;

                startTime = DateTime.Now;

                // now we have all of our triangles assigned to the correct mesh, we can make all of the meshes.
                foreach (MeshData meshdata in meshSectors.Values)
                {
                    // Tell the meshdata class to generate its mesh.
                    meshdata.Commit();
                    // And create a surface object using the mesh.
                    AddSurfaceObject(meshdata.MeshObject, string.Format("SurfaceUnderstanding Mesh-{0}", transform.childCount), transform).AddComponent<WorldAnchor>();

                    // Limit our run time so that we don't cause too many frame drops.
                    if ((DateTime.Now - startTime).TotalMilliseconds > MaxFrameTime)
                    {
                        yield return null;
                        startTime = DateTime.Now;
                    }
                }
            }

            // Wait a frame
            yield return null;
```
to

```
252               yield return null;
                startTime = DateTime.Now;

                // now we have all of our triangles assigned to the correct mesh, we can make all of the meshes.
                foreach (MeshData meshdata in meshSectors.Values)
                {
                    // Tell the meshdata class to generate its mesh.
                    meshdata.Commit();
                    
                    // Limit our run time so that we don't cause too many frame drops.
                    if ((DateTime.Now - startTime).TotalMilliseconds > MaxFrameTime)
                    {
                        yield return null;
                        startTime = DateTime.Now;
                    }
                }
                yield return null;
                // Clear our old surface objects.
                Cleanup();

                foreach (MeshData meshdata in meshSectors.Values)
                {
                    // And create a surface object using the mesh.
                    AddSurfaceObject(meshdata.MeshObject, string.Format("SurfaceUnderstanding Mesh-{0}", transform.childCount), transform).AddComponent<WorldAnchor>();

                }
            }


```

This change gets rid of the frame drop outs at the price of a slight frame-rate reduction on the update beat.   What it is doing is chopping the code up a little bit and ensuring that no yields happen after the call to CleanUp();

I also should note that the finalized mesh (in the original checked in code) was still only getting 29fps even when I am only looking at a small piece of the mesh.   ~~I suspect this can be addressed in the shader by not using a geometry shader~~(EDIT: i tried this, still at 29fps), although it leaves me questioning the effectiveness of this new approach in terms of frame rate improvement.

However, this code does seem to address bug #200 in practice, however this code still does not actually explicitly guarantee one of the meshes wont be too large.

### PR DESCRIPTION
**Pull request by darax**
Fixes #264 
Fixes #207

### CODE REVIEW

**Comment by izatar** [comment]
I just tried this out, and the flickering issue is fixed.
However, I did notice two new things I have not seen before.
First, the button panel that is put up is under the finalized mesh, so I can't see it.  It appears to be too close the mesh plane and the z-ordering is inaccurate.  
The second is that the finalized mesh is not 'water-tight' anymore, there are gaps that have no polygons.

**Comment by izatar** [comment]
20161006_231535_hololens

The attached image shows an example of the gaps in the finalized mesh.

**Comment by izatar** [comment]
After some false starts (like normal calculation), I did find the change that seems to cause this problem:
```
 if (GetComponent<WorldAnchor>() == null)
            {
                gameObject.AddComponent<WorldAnchor>();
            }
```
I don't understand why.  But, there it is.

**Comment by darax** [comment]
I reverted to having an anchor per mesh rather than one anchor.  That seems to have fixed the seams.

**Comment by darax** [comment]
I've also added a change to get rid of the unreachable code warning.

**Comment by izatar** [comment]
This is better, but there are still some seams.
Converting:
`AddSurfaceObject(null, string.Format("SurfaceUnderstanding Mesh-{0}", meshSectorsIndex), transform).AddComponent<WorldAnchor>();`

to
                      `  AddSurfaceObject(null, string.Format("SurfaceUnderstanding Mesh-{0}", meshSectorsIndex), transform);`

seems to remedy this situation.  That is to say, just don't do the worldanchor at all.  It works without them (for me anyway), so I'm not sure if they are necessary.

**Comment by darax** [comment]
We are ignoring the normals that come from the native layer.  I can refactor to use them to see if it helps.  

Regarding the anchor, if you don't have an anchor and lose tracking the meshes will likely be off the surfaces after tracking is restored.

**Comment by darax** [comment]
The normals from the native layer seem to be all zero.   However, in doing so I was forced to remove an optimization where a single vert could be reused multiple times, and I think that will fix the normal problem.  Let me clean this up and I'll push it so that you can try it. 

Thank you very much for the time you spent on this!

**Comment by darax** [comment]
I think I got it. :)  

I tested with a shader that rendered raw normals, and they look consistent, even at the edges (I saw how they weren't before).  I had to stick with a single world anchor or the seams would appear when rendering with wireframes.

**Comment by izatar** [comment]
This is looking good to me.  I did not notice artifacts on the wire mesh.

File: Assets/HoloToolkit/SpatialUnderstanding/Scripts/SpatialUnderstandingCustomMesh.cs [hunk 1]
```diff
@@ -207,6 +236,7 @@ public IEnumerator Import_UnderstandingMesh()
                 (meshIndices != null) &&
                 (meshIndices.Length > 0))
             {
+
```
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/SpatialUnderstanding/Scripts/SpatialUnderstandingCustomMesh.cs]: Nit: extra space

File: Assets/HoloToolkit/SpatialUnderstanding/Scripts/SpatialUnderstandingCustomMesh.cs [hunk 2]
```diff
@@ -6,6 +6,7 @@
 using System;
 using System.Collections.Generic;
 using UnityEngine.VR.WSA;
+using System.Linq;
```
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/SpatialUnderstanding/Scripts/SpatialUnderstandingCustomMesh.cs]: Are we actually using Linq in this class?  If so, do we _need_ to?

**Comment by darax** [comment]
I'm going to merge this now.

**Pull request merged**